portal-web/test/functional/com/liferay/portalweb/macros/LiferayEvent.macro

```diff
@@ -0,0 +1,105 @@
+definition {
+
+	macro initializeLiferayEventLog {
+		var javascript = '''
+const TRACKED_EVENTS = [
+	'addPortlet',
+	'allPortletsReady',
+	'allUploadsComplete',
+	'beforeNavigate',
+	'beforeScreenFlip',
+	'closePortlet',
+	'dataLayoutConfigReady',
+	'destroyPortlet',
+	'editorAPIReady',
+	'endNavigate',
+	'form:registered',
+	'formNavigator:reveal',
+	'dropdownShow',
+	'entitySelectionRemoved',
+	'exitNavigation',
+	'hideNavigationMenu',
+	'initLayout',
+	'inputLocalized:localeChanged',
+	'modalIframeLoaded',
+	'noticeShow',
+	'noticeHide',
+	'popupReady',
+	'portletDragStart',
+	'portletMoved',
+	'portletReady',
+	'ratings:vote',
+	'screenActivate',
+	'screenDeactivate',
+	'screenFlip',
+	'screenLoad',
+	'search-container:registered',
+	'sessionExpired',
+	'showNavigationMenu',
+	'spaRequestTimeout',
+	'startNavigate',
+	'submitForm',
+	'tempFileRemoved',
+	'toggleControls',
+	'updatedLayout',
+];
+
+let __LFR_TRACKED_EVENTS__ = {};
+
+Liferay.on(TRACKED_EVENTS, (event) => {
+	const eventLog = __LFR_TRACKED_EVENTS__[event.type]
+		? [...__LFR_TRACKED_EVENTS__[event.type], event]
+		: [event];
```

Comment:
Thought for the future. At the moment we're storing a reference to the event, but if we ever start using an event system that uses event pooling (ie. recycling event objects to reduce GC pressure; React is an example of system that does this, for example) then we'll probably want to make a "deep" copy of each event and store that instead of a reference. (Hard to do robustly, because at the moment we allow totally arbitrary data to be attached to an event.)

Reply:
Yeah... maybe when we're done with `Modals` we can move to our event system... 😂